File: downloadInvoices.js
/**
 * Nexudus Invoices → ETL Excel Copier → S3 Versioned Upload
 * ---------------------------------------------------------
 * Reads a template, replaces sheet with Nexudus data,
 * saves locally with timestamped filename, then uploads to S3 with versioning.
 * 
 * Creates a new file each run: ETL_MarioDemo_YYYY-MM-DD_HHmmss.xlsx
 * Fetches the previous day's invoice data.
 * Designed to run daily via GitHub Actions workflow.
 */
require("dotenv").config();
const axios = require("axios");
const fs = require("fs-extra");
const path = require("path");
const XLSX = require("xlsx");
const { S3Client, PutObjectCommand } = require("@aws-sdk/client-s3");

// --------------------------------------------------
// CONFIG
// --------------------------------------------------
const TEMPLATE_FILE = path.join(__dirname, "template/ETL_MarioDemo.xlsx");
const OUTPUT_DIR = path.join(__dirname, "output");
const DEST_SHEET_NAME = "Membership invoices";
const NEXUDUS_REPORT_URL = "https://reports.nexudus.com/ReportCenter/Invoices";
const NEXUDUS_TOKEN_URL = "https://spaces.nexudus.com/api/token";

// S3
const S3_BUCKET = process.env.S3_BUCKET || "level39-etl-mario";
const S3_KEY_PREFIX = "output/"; // Files will be saved as: output/ETL_MarioDemo_YYYY-MM-DD_HHmmss.xlsx

const s3Client = new S3Client({
  region: process.env.AWS_REGION || "us-east-1"
});

// --------------------------------------------------
// Generate timestamped filename
// --------------------------------------------------
function generateOutputFilename() {
  const now = new Date();
  const year = now.getUTCFullYear();
  const month = String(now.getUTCMonth() + 1).padStart(2, "0");
  const day = String(now.getUTCDate()).padStart(2, "0");
  const hours = String(now.getUTCHours()).padStart(2, "0");
  const minutes = String(now.getUTCMinutes()).padStart(2, "0");
  const seconds = String(now.getUTCSeconds()).padStart(2, "0");
  
  const timestamp = `${year}-${month}-${day}_${hours}${minutes}${seconds}`;
  const filename = `ETL_MarioDemo_${timestamp}.xlsx`;
  
  return {
    local: path.join(OUTPUT_DIR, filename),
    s3: `${S3_KEY_PREFIX}${filename}`,
    displayName: filename
  };
}

// --------------------------------------------------
// Calculate yesterday's date range (00:00:00 - 23:59:59 UTC)
// --------------------------------------------------
function getDateRange() {
  const yesterday = new Date();
  yesterday.setUTCDate(yesterday.getUTCDate() - 1);

  const startDate = new Date(yesterday);
  startDate.setUTCHours(0, 0, 0, 0);

  const endDate = new Date(yesterday);
  endDate.setUTCHours(23, 59, 59, 999);

  return {
    start: startDate.toISOString().split(".")[0], // 2026-02-11T00:00:00
    end: endDate.toISOString().split(".")[0]      // 2026-02-11T23:59:59
  };
}

// --------------------------------------------------
// Get Nexudus Access Token
// --------------------------------------------------
async function getNexudusToken() {
  try {
    console.log("🔑 Requesting Nexudus authentication token...");
    
    const response = await axios.post(
      NEXUDUS_TOKEN_URL,
      new URLSearchParams({
        grant_type: "password",
        username: process.env.NEXUDUS_USERNAME,
        password: process.env.NEXUDUS_PASSWORD
      }).toString(),
      {
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
          Accept: "application/json"
        },
        timeout: 10000
      }
    );

    if (!response.data?.access_token) {
      throw new Error("No access token received from Nexudus");
    }

    console.log("✔ Nexudus token obtained successfully");
    return response.data.access_token;
  } catch (err) {
    if (err.response?.status === 401) {
      throw new Error("Nexudus authentication failed - invalid credentials");
    } else if (err.code === "ECONNREFUSED" || err.code === "ETIMEDOUT") {
      throw new Error("Cannot reach Nexudus API - network issue or service unavailable");
    }
    throw new Error(`Nexudus token request failed: ${err.message}`);
  }
}

// --------------------------------------------------
// Download Nexudus report
// --------------------------------------------------
async function downloadReport(token) {
  try {
    const { start, end } = getDateRange();
    
    console.log(`📅 Fetching invoices from ${start} to ${end}`);

    const response = await axios.get(NEXUDUS_REPORT_URL, {
      params: {
        businessId: process.env.BUSINESS_ID,
        reportName: "Invoices/InvoicesAccount",
        start: start,
        end: end,
        format: "Excel",
        portrait: false,
        rnd: new Date().toISOString()
      },
      responseType: "arraybuffer",
      headers: {
        Authorization: `Bearer ${token}`,
        Accept: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
      },
      timeout: 30000
    });

    if (!response.data || response.data.length === 0) {
      throw new Error("Nexudus report returned empty data");
    }

    console.log(`✔ Report downloaded successfully (${response.data.length} bytes)`);
    return response.data;
  } catch (err) {
    if (err.response?.status === 404) {
      throw new Error("Nexudus report endpoint not found - check API URL");
    } else if (err.response?.status === 403) {
      throw new Error("Access denied to Nexudus report - check business ID and permissions");
    } else if (err.code === "ETIMEDOUT") {
      throw new Error("Nexudus report request timed out");
    }
    throw new Error(`Failed to download Nexudus report: ${err.message}`);
  }
}

// --------------------------------------------------
// Write report to template & save locally
// --------------------------------------------------
async function writeToTemplate(excelBuffer, outputFile) {
  try {
    await fs.ensureDir(OUTPUT_DIR);
    
    console.log("📝 Processing Excel data...");

    // Load Nexudus workbook
    const sourceWorkbook = XLSX.read(excelBuffer, { type: "buffer" });
    
    if (!sourceWorkbook.SheetNames || sourceWorkbook.SheetNames.length === 0) {
      throw new Error("Nexudus workbook contains no sheets");
    }

    const sourceSheetName = sourceWorkbook.SheetNames[0];
    const sourceSheet = sourceWorkbook.Sheets[sourceSheetName];

    if (!sourceSheet) {
      throw new Error(`Cannot access sheet "${sourceSheetName}" in Nexudus workbook`);
    }

    console.log(`  ├─ Source sheet: "${sourceSheetName}" (${Object.keys(sourceSheet).length} cells)`);

    // Load template workbook
    if (!fs.existsSync(TEMPLATE_FILE)) {
      throw new Error(`Template file not found: ${TEMPLATE_FILE}`);
    }

    const destWorkbook = XLSX.readFile(TEMPLATE_FILE);
    console.log(`  ├─ Template loaded with sheets: ${destWorkbook.SheetNames.join(", ")}`);

    // Remove existing sheet if exists
    if (destWorkbook.SheetNames.includes(DEST_SHEET_NAME)) {
      console.log(`  ├─ Removing existing sheet: "${DEST_SHEET_NAME}"`);
      delete destWorkbook.Sheets[DEST_SHEET_NAME];
      destWorkbook.SheetNames = destWorkbook.SheetNames.filter(
        name => name !== DEST_SHEET_NAME
      );
    }

    // Append fresh sheet
    XLSX.utils.book_append_sheet(destWorkbook, sourceSheet, DEST_SHEET_NAME);
    console.log(`  └─ New sheet added: "${DEST_SHEET_NAME}"`);

    // Save locally with timestamped filename
    XLSX.writeFile(destWorkbook, outputFile);
    console.log(`✔ Local ETL file saved: ${outputFile}`);
  } catch (err) {
    throw new Error(`Failed to process Excel template: ${err.message}`);
  }
}

// --------------------------------------------------
// Upload to S3 (versioned)
// --------------------------------------------------
async function uploadToS3(localFile, s3Key, displayName) {
  try {
    console.log("☁️  Uploading to S3...");

    const fileContent = fs.readFileSync(localFile);

    const result = await s3Client.send(
      new PutObjectCommand({
        Bucket: S3_BUCKET,
        Key: s3Key,
        Body: fileContent
      })
    );

    console.log(`✔ File uploaded to S3: s3://${S3_BUCKET}/${s3Key}`);
    if (result.VersionId) {
      console.log(`📦 Version ID: ${result.VersionId}`);
    } else {
      console.log("⚠️  Note: S3 versioning may not be enabled on this bucket");
    }
  } catch (err) {
    if (err.code === "NoSuchBucket") {
      throw new Error(`S3 bucket not found: ${S3_BUCKET}`);
    } else if (err.code === "AccessDenied") {
      throw new Error("Access denied to S3 bucket - check AWS credentials and permissions");
    }
    throw new Error(`Failed to upload to S3: ${err.message}`);
  }
}

// --------------------------------------------------
// Main ETL Job
// --------------------------------------------------
async function runETL() {
  try {
    console.log("=".repeat(50));
    console.log("🚀 Starting Nexudus ETL Job");
    console.log("=".repeat(50));
    console.log(`⏰ Run time: ${new Date().toISOString()}`);
    console.log();

    // Generate timestamped output filenames
    const fileNames = generateOutputFilename();
    console.log(`📁 Output file: ${fileNames.displayName}`);
    console.log();

    // Validate environment variables
    const requiredEnvVars = [
      "NEXUDUS_USERNAME",
      "NEXUDUS_PASSWORD",
      "BUSINESS_ID",
      "AWS_ACCESS_KEY_ID",
      "AWS_SECRET_ACCESS_KEY"
    ];

    const missingEnvVars = requiredEnvVars.filter(envVar => !process.env[envVar]);
    if (missingEnvVars.length > 0) {
      throw new Error(`Missing required environment variables: ${missingEnvVars.join(", ")}`);
    }

    // Execute ETL pipeline
    const token = await getNexudusToken();
    const reportBuffer = await downloadReport(token);
    await writeToTemplate(reportBuffer, fileNames.local);
    await uploadToS3(fileNames.local, fileNames.s3, fileNames.displayName);

    console.log();
    console.log("=".repeat(50));
    console.log("✅ ETL Job Completed Successfully");
    console.log("=".repeat(50));
  } catch (err) {
    console.error();
    console.error("=".repeat(50));
    console.error("❌ ETL Job Failed");
    console.error("=".repeat(50));
    console.error(`⏰ Error time: ${new Date().toISOString()}`);
    console.error(`📌 Error: ${err.message}`);
    console.error("=".repeat(50));
    
    // Log full stack trace for debugging
    if (process.env.DEBUG === "true") {
      console.error("\nFull stack trace:");
      console.error(err.stack);
    }

    // Exit with error code for GitHub Actions to detect failure
    process.exit(1);
  }
}

// --------------------------------------------------
// Run ETL
// --------------------------------------------------
runETL();

File: downloadInvoices.py
#!/usr/bin/env python3
"""
Nexudus Invoices → ETL Excel Copier → S3 Versioned Upload
---------------------------------------------------------
Opens template, adds Nexudus data to Membership Invoices sheet, saves and uploads.
Simple approach: minimal changes to preserve all Excel features.
"""

import os
import sys
from datetime import datetime, timedelta
import requests
from io import BytesIO
import json
from openpyxl import load_workbook
from openpyxl.utils import get_column_letter
import boto3
from botocore.exceptions import ClientError

# --------------------------------------------------
# CONFIG
# --------------------------------------------------
TEMPLATE_FILE = "template/ETL_MarioDemo.xlsx"
OUTPUT_DIR = "output"
DEST_SHEET_NAME = "Membership invoices"
DATA_START_ROW = 1  # Data starts at row 1

NEXUDUS_REPORT_URL = "https://reports.nexudus.com/ReportCenter/Invoices"
NEXUDUS_TOKEN_URL = "https://spaces.nexudus.com/api/token"

# S3
S3_BUCKET = os.getenv("S3_BUCKET", "level39-etl-mario")
S3_KEY_PREFIX = "output/"

# Create output directory if it doesn't exist
os.makedirs(OUTPUT_DIR, exist_ok=True)

# --------------------------------------------------
# Generate timestamped filename
# --------------------------------------------------
def generate_output_filename():
    now = datetime.utcnow()
    timestamp = now.strftime("%Y-%m-%d_%H%M%S")
    filename = f"ETL_MarioDemo_{timestamp}.xlsx"
    
    return {
        "local": os.path.join(OUTPUT_DIR, filename),
        "s3": f"{S3_KEY_PREFIX}{filename}",
        "displayName": filename
    }

# --------------------------------------------------
# Calculate yesterday's date range
# --------------------------------------------------
def get_date_range():
    yesterday = datetime.utcnow() - timedelta(days=1)
    
    start_date = datetime(yesterday.year, yesterday.month, yesterday.day, 0, 0, 0)
    end_date = datetime(yesterday.year, yesterday.month, yesterday.day, 23, 59, 59)
    
    return {
        "start": start_date.isoformat(),
        "end": end_date.isoformat()
    }

# --------------------------------------------------
# Get Nexudus Access Token
# --------------------------------------------------
def get_nexudus_token():
    try:
        print("🔑 Requesting Nexudus authentication token...")
        
        response = requests.post(
            NEXUDUS_TOKEN_URL,
            data={
                "grant_type": "password",
                "username": os.getenv("NEXUDUS_USERNAME"),
                "password": os.getenv("NEXUDUS_PASSWORD")
            },
            headers={
                "Content-Type": "application/x-www-form-urlencoded",
                "Accept": "application/json"
            },
            timeout=10
        )
        
        response.raise_for_status()
        
        data = response.json()
        if "access_token" not in data:
            raise Exception("No access token received from Nexudus")
        
        print("✔ Nexudus token obtained successfully")
        return data["access_token"]
        
    except requests.exceptions.RequestException as err:
        if "401" in str(err):
            raise Exception("Nexudus authentication failed - invalid credentials")
        elif "timeout" in str(err).lower():
            raise Exception("Cannot reach Nexudus API - network issue or service unavailable")
        raise Exception(f"Nexudus token request failed: {str(err)}")

# --------------------------------------------------
# Download Nexudus report
# --------------------------------------------------
def download_report(token):
    try:
        date_range = get_date_range()
        start = date_range["start"].split(".")[0]
        end = date_range["end"].split(".")[0]
        
        print(f"📅 Fetching invoices from {start} to {end}")
        
        response = requests.get(
            NEXUDUS_REPORT_URL,
            params={
                "businessId": os.getenv("BUSINESS_ID"),
                "reportName": "Invoices/InvoicesAccount",
                "start": start,
                "end": end,
                "format": "Excel",
                "portrait": "false",
                "rnd": datetime.utcnow().isoformat()
            },
            headers={
                "Authorization": f"Bearer {token}",
                "Accept": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
            },
            timeout=30
        )
        
        response.raise_for_status()
        
        if len(response.content) == 0:
            raise Exception("Nexudus report returned empty data")
        
        print(f"✔ Report downloaded successfully ({len(response.content)} bytes)")
        return BytesIO(response.content)
        
    except requests.exceptions.RequestException as err:
        if "404" in str(err):
            raise Exception("Nexudus report endpoint not found - check API URL")
        elif "403" in str(err):
            raise Exception("Access denied to Nexudus report - check business ID and permissions")
        elif "timeout" in str(err).lower():
            raise Exception("Nexudus report request timed out")
        raise Exception(f"Failed to download Nexudus report: {str(err)}")

# --------------------------------------------------
# Update template with data
# --------------------------------------------------
def update_template_with_data(report_buffer, output_file):
    try:
        print("📝 Processing Excel data...")
        
        # Load the Nexudus report
        print("  ├─ Loading Nexudus report...")
        report_wb = load_workbook(report_buffer)
        
        if not report_wb.sheetnames:
            raise Exception("Nexudus workbook contains no sheets")
        
        print(f"  ├─ Available sheets: {', '.join(report_wb.sheetnames)}")
        
        # Find sheet with most data
        source_sheet_name = report_wb.sheetnames[0]
        max_cells = 0
        
        for sheet_name in report_wb.sheetnames:
            sheet = report_wb[sheet_name]
            cell_count = sum(1 for row in sheet.iter_rows() for cell in row if cell.value is not None)
            print(f"  ├─ '{sheet_name}': {cell_count} cells")
            
            if cell_count > max_cells:
                max_cells = cell_count
                source_sheet_name = sheet_name
        
        print(f"  ├─ Using sheet: '{source_sheet_name}' ({max_cells} cells)")
        
        source_sheet = report_wb[source_sheet_name]
        
        # Load template
        print("  ├─ Loading template...")
        if not os.path.exists(TEMPLATE_FILE):
            raise Exception(f"Template file not found: {TEMPLATE_FILE}")
        
        template_wb = load_workbook(TEMPLATE_FILE)
        print(f"  ├─ Template sheets: {', '.join(template_wb.sheetnames)}")
        
        if DEST_SHEET_NAME not in template_wb.sheetnames:
            raise Exception(f"Sheet '{DEST_SHEET_NAME}' not found in template")
        
        dest_sheet = template_wb[DEST_SHEET_NAME]
        print(f"  ├─ Found destination sheet: '{DEST_SHEET_NAME}'")
        
        # Copy data from source to destination starting at row 6
        print(f"  ├─ Copying data starting at row {DATA_START_ROW}...")
        
        copied_rows = 0
        copied_cells = 0
        
        # Iterate through source sheet
        for src_row_idx, src_row in enumerate(source_sheet.iter_rows()):
            dest_row_idx = DATA_START_ROW + src_row_idx
            
            for src_col_idx, src_cell in enumerate(src_row, start=1):
                if src_cell.value is not None:
                    dest_cell = dest_sheet.cell(row=dest_row_idx, column=src_col_idx)
                    
                    # Copy value
                    dest_cell.value = src_cell.value
                    
                    # Copy data type if available
                    if hasattr(src_cell, 'data_type'):
                        dest_cell.data_type = src_cell.data_type
                    
                    # Copy number format if available
                    if src_cell.number_format:
                        dest_cell.number_format = src_cell.number_format
                    
                    copied_cells += 1
            
            copied_rows += 1
        
        print(f"  ├─ Copied {copied_rows} rows ({copied_cells} cells)")
        print(f"  └─ Saving template...")
        
        # Save to output file
        template_wb.save(output_file)
        print(f"✔ Local ETL file saved: {output_file}")
        
    except Exception as err:
        raise Exception(f"Failed to process Excel template: {str(err)}")

# --------------------------------------------------
# Upload to S3
# --------------------------------------------------
def upload_to_s3(local_file, s3_key, display_name):
    try:
        print("☁️  Uploading to S3...")
        
        s3_client = boto3.client(
            "s3",
            region_name=os.getenv("AWS_REGION", "us-east-1"),
            aws_access_key_id=os.getenv("AWS_ACCESS_KEY_ID"),
            aws_secret_access_key=os.getenv("AWS_SECRET_ACCESS_KEY")
        )
        
        with open(local_file, "rb") as f:
            response = s3_client.put_object(
                Bucket=S3_BUCKET,
                Key=s3_key,
                Body=f.read()
            )
        
        print(f"✔ File uploaded to S3: s3://{S3_BUCKET}/{s3_key}")
        
        if "VersionId" in response:
            print(f"📦 Version ID: {response['VersionId']}")
        else:
            print("⚠️  Note: S3 versioning may not be enabled on this bucket")
            
    except ClientError as err:
        if err.response['Error']['Code'] == 'NoSuchBucket':
            raise Exception(f"S3 bucket not found: {S3_BUCKET}")
        elif err.response['Error']['Code'] == 'AccessDenied':
            raise Exception("Access denied to S3 bucket - check AWS credentials and permissions")
        raise Exception(f"Failed to upload to S3: {str(err)}")

# --------------------------------------------------
# Main ETL Job
# --------------------------------------------------
def run_etl():
    try:
        print("=" * 50)
        print("🚀 Starting Nexudus ETL Job")
        print("=" * 50)
        print(f"⏰ Run time: {datetime.utcnow().isoformat()}")
        print()
        
        file_names = generate_output_filename()
        print(f"📁 Output file: {file_names['displayName']}")
        print()
        
        # Validate environment variables
        required_vars = [
            "NEXUDUS_USERNAME",
            "NEXUDUS_PASSWORD",
            "BUSINESS_ID",
            "AWS_ACCESS_KEY_ID",
            "AWS_SECRET_ACCESS_KEY"
        ]
        
        missing_vars = [var for var in required_vars if not os.getenv(var)]
        if missing_vars:
            raise Exception(f"Missing required environment variables: {', '.join(missing_vars)}")
        
        # Execute ETL pipeline
        token = get_nexudus_token()
        report_buffer = download_report(token)
        update_template_with_data(report_buffer, file_names["local"])
        upload_to_s3(file_names["local"], file_names["s3"], file_names["displayName"])
        
        print()
        print("=" * 50)
        print("✅ ETL Job Completed Successfully")
        print("=" * 50)
        
    except Exception as err:
        print()
        print("=" * 50)
        print("❌ ETL Job Failed")
        print("=" * 50)
        print(f"⏰ Error time: {datetime.utcnow().isoformat()}")
        print(f"📌 Error: {str(err)}")
        print("=" * 50)
        
        if os.getenv("DEBUG") == "true":
            import traceback
            print("\nFull stack trace:")
            traceback.print_exc()
        
        sys.exit(1)

# --------------------------------------------------
# Run ETL
# --------------------------------------------------
if __name__ == "__main__":
    run_etl()
